### PR TITLE
Kafka connect sink for v2

### PR DESCRIPTION
**Pull request by Akeboshiwind**
A port of the kafka connect sink from v1 to v2

TODO:
- [x] Figure out distribution
- [ ] Add to docs
- [x] Figure out where to put my local testing docker-compose etc

### CODE REVIEW

File: .github/workflows/build-connect.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: XTDB Kafka Connect
+run-name: XTDB Kafka Connect
+
+on:
+  push:
+    branches:
+      - kafka-connect-latest
+
+jobs:
+  deploy-standalone-docker:
```
Comment (jarohen) [inline .github/workflows/build-connect.yml]: standalone-docker?

File: modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+(defn- get-struct-contents [val]
+  (cond
+    (instance? Struct val)
+    (let [struct-schema (.schema ^Struct val)
+          struct-fields (.fields ^Schema struct-schema)]
+      (reduce conj
+              (map (fn [^Field field] {(keyword (.name field)) (get-struct-contents (.get ^Struct val field))})
+                   struct-fields)))
+    (instance? java.util.ArrayList val) (into [] (map get-struct-contents val))
```
Comment (jarohen) [inline modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj]: would `List` suffice here, save us being over-specific on the concrete type?

File: modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    (instance? java.util.HashMap val) (zipmap (map keyword (.keySet ^java.util.HashMap val)) (map get-struct-contents (.values ^java.util.HashMap val)))
```
Comment (jarohen) [inline modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj]: `Map`, likewise

File: modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj (hunk 2 above)
Comment (Akeboshiwind) [reply modules/kafka-connect/src/main/clojure/xtdb/kafka/connect.clj]: These are just copy paste from the [v1 implementation](https://github.com/xtdb/xtdb/blob/06132bc1fad63944d144d79a90053b05d78a41c8/modules/kafka-connect/src/xtdb/kafka/connect.clj#L33).
Happy to change though

**Comment by Akeboshiwind** [comment]
While testing this out I had a setup to run everything locally. Including in the PR for future reference.

<details><summary>`docker-compose.yml`</summary>

```yml
---
services:
  zookeeper:
    image: confluentinc/cp-zookeeper:7.7.0
    container_name: zookeeper
    environment:
      ZOOKEEPER_CLIENT_PORT: 2181
      ZOOKEEPER_TICK_TIME: 2000

  kafka:
    image: confluentinc/cp-kafka:7.7.0
    container_name: kafka
    depends_on:
      - zookeeper
    ports:
      - 29092:29092
    environment:
      KAFKA_BROKER_ID: 1
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092,PLAINTEXT_HOST://localhost:29092
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
      KAFKA_INTER_BROKER_LISTENER_NAME: PLAINTEXT
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1

  kafka-connect:
    image: confluentinc/cp-kafka-connect:7.7.0
    container_name: kafka-connect
    depends_on:
      - kafka
    ports:
      - 28083:8083
    volumes:
      - $PWD/data:/data
    environment:
      CONNECT_BOOTSTRAP_SERVERS: kafka:9092
      CONNECT_REST_ADVERTISED_HOST_NAME: localhost
      CONNECT_REST_PORT: 8083
      CONNECT_GROUP_ID: quickstart
      CONNECT_CONFIG_STORAGE_TOPIC: quickstart-config
      CONNECT_OFFSET_STORAGE_TOPIC: quickstart-offsets
      CONNECT_STATUS_STORAGE_TOPIC: quickstart-status
      CONNECT_KEY_CONVERTER: org.apache.kafka.connect.json.JsonConverter
      CONNECT_VALUE_CONVERTER: org.apache.kafka.connect.json.JsonConverter
      CONNECT_KEY_CONVERTER_SCHEMAS_ENABLE: true
      CONNECT_VALUE_CONVERTER_SCHEMAS_ENABLE: true
      CONNECT_PLUGIN_PATH: /usr/share/java,/data/connect-jars
      CONNECT_CONFIG_STORAGE_REPLICATION_FACTOR: 1
      CONNECT_OFFSET_STORAGE_REPLICATION_FACTOR: 1
      CONNECT_STATUS_STORAGE_REPLICATION_FACTOR: 1

  xtdb:
    container_name: xtdb
    image: ghcr.io/xtdb/xtdb-standalone-ea:latest
    ports:
      # HTTP port for local testing
      - 6543:3000
```

</details>

<details><summary>`connect` (bash script)</summary>

```bash
#!/usr/bin/env bash


# >> Config

connect_url="http://localhost:28083"
connector_name="xtdb-sink-devices"
bootstrap_server="localhost:29092"



# >> Utils

script_path=$(dirname $(readlink -f $0))

info () { echo "[INFO] $1"; }

err () { echo "[ERROR] $1"; }

has_jq() {
    command -v jq > /dev/null
}

format_json() {
    if has_jq; then
        jq . $1
    else
        cat $1
    fi
}



# >> Commands

## Connect

connect_status() {
    curl -s -XGET $connect_url | format_json
}

connect_logs() {
    docker-compose -f $script_path/docker-compose.yml logs $@
}

connect_plugins() {
    curl -s -XGET $connect_url/connector-plugins | format_json
}

connect_up() {
    docker-compose -f $script_path/docker-compose.yml up -d
}

connect_down() {
    docker-compose -f $script_path/docker-compose.yml down
}

## Connector

has_connector() {
    curl -s $connect_url/connectors | grep -q "\"$connector_name\""
}

upload_connector() {
    curl -s \
        -XPOST \
        -H "Content-Type: application/json" \
        -d "@$script_path/connector.json" \
        $connect_url/connectors \
        | format_json
}

connector_status() {
    curl -s \
        -XGET \
        $connect_url/connectors/$connector_name/status \
        | format_json
}

connector_get_config() {
    curl -s \
        -XGET \
        $connect_url/connectors/$connector_name/config \
        | format_json
}

resume_connector() {
    curl -XPUT $connect_url/connectors/$connector_name/resume
}

stop_connector() {
    curl -XPUT $connect_url/connectors/$connector_name/stop
}

delete_connector() {
    curl -XDELETE $connect_url/connectors/$connector_name
}

## Connectors

upload_connectors() {
    for config in $(cat $script_path/connectors.json | jq -c '.[]'); do
        curl -s \
          -XPOST \
          -H "Content-Type: application/json" \
          -d "$config" \
          $connect_url/connectors \
          | format_json
    done
}

connectors_status() {
    for name in $(cat $script_path/connectors.json | jq -cr '.[].name'); do
        curl -s \
            -XGET \
            $connect_url/connectors/$name/status \
            | format_json
    done
}

delete_connector() {
    for name in $(cat $script_path/connectors.json | jq -cr '.[].name'); do
        curl -XDELETE $connect_url/connectors/$name
    done
}

## Data

random_measurement() {
    echo $RANDOM
}

# `xt/id` in the value
struct_value_data() {
    batches=${1:-1}
    max_id=${2:-10}
    for _ in $(seq 1 $batches); do
        for id in $(seq 1 $max_id); do
            echo '{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" }, { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" } ] }, "payload":{"xt/id":'$id', "metric":"Humidity", "measurement":'$(random_measurement)'} }'
        done
    done
}

# id as primitive key
primitive_keyed_data() {
    batches=${1:-1}
    max_id=${2:-10}
    for _ in $(seq 1 $batches); do
        for id in $(seq 1 $max_id); do
            echo $id'|{ "schema": { "type":"struct", "fields":[ { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" } ] }, "payload":{"metric":"Humidity", "measurement":'$(random_measurement)'} }'
        done
    done
}

# id as struct key
struct_keyed_data() {
    batches=${1:-1}
    max_id=${2:-10}
    for _ in $(seq 1 $batches); do
        for id in $(seq 1 $max_id); do
            echo '{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" } ] }, "payload":{"xt/id":'$id'} }|{ "schema": { "type":"struct", "fields":[ { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" } ] }, "payload":{"metric":"Humidity", "measurement":'$(random_measurement)'} }'
        done
    done
}

# Some manually crafted data
valid_time_readings_data() {
    cat << EOF
{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" }, { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" }, { "type":"string", "optional": false, "field": "validFrom"}, { "type":"string", "optional": false, "field": "validTo"} ] }, "payload":{"xt/id":1, "metric":"Humidity", "validFrom":"2024-01-01T00:00:00Z", "validTo":"2024-01-01T00:05:00Z", "measurement":$(random_measurement)} }
{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" }, { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" }, { "type":"string", "optional": false, "field": "validFrom"}, { "type":"string", "optional": false, "field": "validTo"} ] }, "payload":{"xt/id":1, "metric":"Humidity", "validFrom":"2024-01-01T00:05:00Z", "validTo":"2024-01-01T00:10:00Z", "measurement":$(random_measurement)} }
{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" }, { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" }, { "type":"string", "optional": false, "field": "validFrom"}, { "type":"string", "optional": false, "field": "validTo"} ] }, "payload":{"xt/id":1, "metric":"Humidity", "validFrom":"2024-01-01T00:10:00Z", "validTo":"2024-01-01T00:15:00Z", "measurement":$(random_measurement)} }
{ "schema": { "type":"struct", "fields":[ { "type":"int64", "optional": false, "field": "xt/id" }, { "type":"string", "optional": false, "field": "metric" }, { "type":"int32", "optional": false, "field": "measurement" }, { "type":"string", "optional": false, "field": "validFrom"}, { "type":"string", "optional": false, "field": "validTo"} ] }, "payload":{"xt/id":1, "metric":"Humidity", "validFrom":"2024-01-01T00:15:00Z", "validTo":"2024-01-01T00:20:00Z", "measurement":$(random_measurement)} }
EOF
}

console_producer() {
    topic=$1
    $script_path/bin/kafka-console-producer.sh \
      --bootstrap-server $bootstrap_server \
      --topic $topic
}

keyed_console_producer() {
    topic=$1
    $script_path/bin/kafka-console-producer.sh \
      --bootstrap-server $bootstrap_server \
      --topic $topic \
      --property "parse.key=true" \
      --property "key.separator=|"
}

upload_data() {
    struct_value_data 100 | console_producer orders
    primitive_keyed_data 100 | keyed_console_producer orders_primitive_key
    struct_keyed_data 100 | keyed_console_producer orders_struct_key
    valid_time_readings_data | console_producer orders_valid_time
}

listen_data() {
    $script_path/bin/kafka-console-consumer.sh \
        --bootstrap-server $bootstrap_server \
        --topic $1
}



# >> Cli

usage_connect() {
    echo "Usage: $0 connect {status|plugins|logs|up|down}"
}

usage_connector() {
    echo "Usage: $0 connector {up|status|status-trace|logs|config|down}"
}

usage_data() {
    echo "Usage: $0 data upload"
    echo "Usage: $0 data listen <topic>"
}

usage() {
    usage_connect
    usage_connector
    usage_data
}

case $1 in
    connect)
        case $2 in
            status)
                connect_status
                ;;
            plugins)
                connect_plugins
                ;;
            logs)
                shift
                shift
                connect_logs $@
                ;;
            up)
                connect_up
                ;;
            down)
                connect_down
                ;;
            *)
                err "Invalid connect subcommand"
                usage_connect
                exit 1
        esac
        ;;
    connector)
        case $2 in
            up)
                if ! has_connector; then
                    info "Updating connector config"
                    upload_connector
                else
                    info "Connector already started"
                fi
                ;;
            status)
                if has_connector; then
                    connector_status
                else
                    info "Connector not started"
                fi
                ;;
            status-trace)
                if has_connector; then
                    if has_jq; then
                        connector_status | jq -r '.tasks[].trace'
                    else
                        err "jq not found"
                    fi
                else
                    info "Connector not started"
                fi
                ;;
            logs)
                shift
                shift
                connect_logs kafka-connect $@
                ;;
            config)
                if has_connector; then
                    connector_get_config
                else
                    info "Connector not started"
                fi
                ;;
            down)
                if has_connector; then
                    info "Deleting connector"
                    delete_connector
                else
                    info "Connector not started"
                fi
                ;;
            *)
                err "Invalid connector subcommand"
                usage_connector
                exit 1
        esac
        ;;
    connectors)
        case $2 in
            up)
                if ! has_connector; then
                    info "Updating connector config"
                    upload_connectors
                else
                    info "Connector already started"
                fi
                ;;
            status)
                if has_connector; then
                    connectors_status
                else
                    info "Connector not started"
                fi
                ;;
            status-trace)
                if has_connector; then
                    if has_jq; then
                        connectors_status | jq -r '.tasks[].trace'
                    else
                        err "jq not found"
                    fi
                else
                    info "Connector not started"
                fi
                ;;
            down)
                if has_connector; then
                    info "Deleting connector"
                    delete_connectors
                else
                    info "Connector not started"
                fi
                ;;
            *)
                err "Invalid connector subcommand"
                usage_connector
                exit 1
        esac
        ;;
    data)
        case $2 in
            upload)
                upload_data
                ;;
            listen)
                if [ -z $3 ]; then
                    err "Missing topic"
                    usage_data
                    exit 1
                else
                    listen_data $3
                fi
                ;;
            *)
                err "Invalid data subcommand"
                usage_data
                exit 1
                ;;
        esac
        ;;
    *)
        err "Invalid command"
        usage
        exit 1
        ;;
esac
```

</details>

<details><summary>`connector.json`</summary>

```json
{
    "name": "xtdb-jdbc-sink-connector",
    "config": {
        "connector.class": "xtdb.kafka.connect.XtdbSinkConnector",
        "url": "http://xtdb:3000",

        "topics": "orders_valid_time",
        "tasks.max": 1,

        "id.mode": "record_value",
        "id.field": "xt/id",

        "validFrom.field": "validFrom",
        "validTo.field": "validTo",

        "table.name.format": "readings",

        "transforms": "convertValidFrom,convertValidTo",

        "transforms.convertValidFrom.type": "org.apache.kafka.connect.transforms.TimestampConverter$Value",
        "transforms.convertValidFrom.field": "validFrom",
        "transforms.convertValidFrom.target.type": "Timestamp",
        "transforms.convertValidFrom.format": "yyyy-MM-dd'T'HH:mm:ssX",

        "transforms.convertValidTo.type": "org.apache.kafka.connect.transforms.TimestampConverter$Value",
        "transforms.convertValidTo.field": "validTo",
        "transforms.convertValidTo.target.type": "Timestamp",
        "transforms.convertValidTo.format": "yyyy-MM-dd'T'HH:mm:ssX",

        "key.converter": "org.apache.kafka.connect.json.JsonConverter",
        "key.converter.schemas.enable": "true",

        "value.converter": "org.apache.kafka.connect.json.JsonConverter",
        "value.converter.schemas.enable": "true"
    }
}
```

</details>

Make sure all these files are in the root directory of a downloaded of the [latest kafka release](https://www.apache.org/dyn/closer.cgi?path=/kafka/3.8.0/kafka_2.13-3.8.0.tgz) (I use the tools in the `bin/` directory).

Make sure to upload the connector to `data/connect-jars/` relative to this same root.

## Workflow

View available commands: `./connect`

Start up the docker containers:
1. `./connect connect up`
2. `./connect connect status` (wait until this returns some json)
    - Alternatively: `watch ./connect connect status`

Start up a connector: `./connect connector up`

Upload some data: `./connect data upload`

Check the status of the connector: `./connect connector status`
If there's an error you can view it more easily with `./connect connector status-trace`

You can also upload multiple connectors using `./connect connectors up`

**Pull request merged**